compiler/resolution/functionResolution.cpp

```diff
@@ -3523,7 +3533,28 @@ FnSymbol* resolveNormalCall(CallExpr* call, bool checkonly) {
      info.call->id == explainCallID);
   DisambiguationContext DC(&info.actuals, scope, explain);
 
-  ResolutionCandidate* best = disambiguateByMatch(candidates, DC);
+  ResolutionCandidate* bestRef = disambiguateByMatch(candidates, DC, FIND_REF);
+  ResolutionCandidate* bestValue = disambiguateByMatch(candidates, DC, FIND_NOT_REF);
+
+  // If one requires more promotion than the other, this is not a ref-pair.
+  if (bestRef && bestValue && isBetterMatch(bestRef, bestValue, DC, true)) {
+    bestValue = NULL; // Don't consider the value function.
+  }
+  if (bestRef && bestValue && isBetterMatch(bestValue, bestRef, DC, true)) {
+    bestRef = NULL; // Don't consider the ref function.
+  }
+
+  ResolutionCandidate* best = bestRef;
+  if (!best) best = bestValue;
+
+  CallExpr* valueCall = NULL;
```

Comment:
I would like this comment right before the 'valueCall' decl:

If we have both ref and value matches:
- 'call' will invoke the ref function best->fn
- 'valueCall' will invoke the value function bestValue->fn
- we will manipulate these two side by side